src/main/java/org/thoughtcrime/securesms/components/emoji/AutoScaledEmojiTextView.java

```diff
@@ -50,36 +50,34 @@ private float getTextScale(String text) {
     if (text.length() > 21) {
       return 1;
     }
-    int emojiCount = countGraphemes(text, 8);
-    if (emojiCount > 8) {
+    int emojiCount = countEmojis(text, 8);
+    if (emojiCount > 8 || emojiCount == 0) {
       return 1;
     }
 
-    boolean isEmoji = emojiRegex.matcher(text).matches();
-    if (isEmoji) {
-      float scale = 1.25f;
-      if (emojiCount <= 6) scale += 0.25f;
-      if (emojiCount <= 4) scale += 0.25f;
-      if (emojiCount <= 2) scale += 0.25f;
-      return scale;
-    }
-    return 1;
+    float scale = 1.25f;
+    if (emojiCount <= 6) scale += 0.25f;
+    if (emojiCount <= 4) scale += 0.25f;
+    if (emojiCount <= 2) scale += 0.25f;
+    return scale;
   }
 
-  public static int countGraphemes(String text, int max) {
-    // Create a BreakIterator for grapheme boundaries
+  /**
+   * Returns the number of emojis if there are only emojis, or 0 otherwise.
+   */
+  public static int countEmojis(String text, int max) {
     BreakIterator graphemeIterator = BreakIterator.getCharacterInstance(Locale.getDefault());
 
-    // Set the text to be analyzed
     graphemeIterator.setText(text);
 
-    // Initialize count
     int graphemeCount = 0;
 
     // Iterate over the text and count graphemes
     int start = graphemeIterator.first();
     for (int end = graphemeIterator.next(); end != BreakIterator.DONE; start = end, end = graphemeIterator.next()) {
-      if (++graphemeCount > max) break;
+      String grapheme = text.substring(start, end);
+      if (!emojiRegex.matcher(grapheme).matches()) return 0;
+      if (++graphemeCount > max) return 0;
```

Comment:
now the returned value is never `> max`  so the `if` condition checking `emojiCount > 8` can be removed, the function doc-string needs also to indicate this. I would prefer returning `-1` in this case then we can check for `emojiCount > 0`

Reply:
did you test the performance impact of regex over the whole text vs doing it for each grapheme? in theory it should be cheap since we do it at most in 8 graphemes and break the loop otherwise

since now this is more robust and we don't loop over whole text, maybe makes sense to increase the length restriction from 21 to 56

there is another PR modifying this, maybe first merge that one and rebase

Reply:
I finally ran some performance tests (4 times opening a chat with a lot of emojis, 2 times opening a chat with just 2 emoji messages). The performance varies a lot, and all in all, it's inconclusive

getTextScale()

Chat with a lot of emojis, new:
16ms, of which 7.4ms in getCharacterInstance, 3.5ms in setText, 4.3ms in next, 1.15ms in matches
3.6ms, of which 1ms in getCharacterInstance, 1ms in setText, 1ms in matches
9ms, of which 5.6ms in getCharacterInstance, 1.5ms in next, 1ms in matcher, 1ms in matches
8.3ms, of which 4.3 in getCharacterInstance, 1.7 in next, 1.4 in setText, 1ms in memcmp (weird...)

Chat with just two messages with emojis, new:
0.8ms, all in setText
1ms, all in next

Chat with a lot of emojis, old:
9ms, of which 5ms in getCharacterInstance, 2ms in next, 1ms in setText, 1ms in matches
1ms, all in matches
5.7ms, of which 4.7 in getCharacterInstance, 1ms in matches

Chat with just two messages with emojis, old:
not present at all
1ms, all in next

Reply:
I do find it a bit surprising that this PR doesn't improve performance; I'd be fine with not merging if we say that performance is more important than enlarging all emojis here